cli/fabric.c

```diff
@@ -1435,6 +1435,105 @@ static int gfms_events(int argc, char **argv)
 	return ret;
 }
 
+static int string_to_dword_data(char *str, unsigned int *dw_data,
+				int *data_len)
+{
+	char *tmp;
+	uint32_t num;
+	char *p = NULL;
+	uint32_t max_len;
+	uint32_t raw_data_len = 0;
+
+	max_len = *data_len;
+	memset(dw_data, 0, max_len);
+
+	p = strtok((char *)str, " ");
+	while(p) {
+		num = strtoul(p, &tmp, 0);
+
+		if (*tmp != '\0')
+			return 1;
+
+		dw_data[raw_data_len] = num;
+
+		raw_data_len++;
+		if(raw_data_len >= max_len)
+			return 1;
+
+		p = strtok(NULL, " ");
+	}
+
+	*data_len = raw_data_len;
+	return 0;
+}
+
+#define CMD_DESC_DEV_MANAGE "send a device specific manage command"
+static int device_manage(int argc, char **argv)
+{
+	int i;
+	int ret;
+	int data_len;
+	char *cmd_string = NULL;
+	struct switchtec_device_manage_rsp rsp;
+
+	static struct {
+		struct switchtec_dev *dev;
+		struct switchtec_device_manage_req req;
+	} cfg = {
+		.req.hdr.pdfid = 0xffff,
+	};
+
+	const struct argconfig_options opts[] = {
+		DEVICE_OPTION,
+		{"pdfid", 'f', "NUM", CFG_INT, &cfg.req.hdr.pdfid,
+		 required_argument, "Endpoint function's FID",
+		 .require_in_usage = 1},
+		{"cmd_data", 'c', "String", CFG_STRING, &cmd_string,
+		 required_argument, .require_in_usage = 1,
+		 .help= "Command raw data in dword, format example: \"0x040b0006 0x00000001\""},
```

Comment:
This is exceptionally gross asking users to pass random data on the command line. switchtec-user was written to get away from that nonsense we saw in its predecessor.

How about writing a proper interface?

Reply:
The data exchanged depends completely on the device, so there is no good way to further structure the data. 

How about something like:
```switchtec fabric device-manage --dw0 0x0000001 --dw1 0x000002 --dw2 0x00003 ...```
This allows user to specify by DWORD instead of a string.

Reply:
Well what actually is the data.... There's no documentation, how are users supposed to figure this out?

Reply:
This command provides a way for user to talk to EP device _directly_. The data sent using this command will be passed to the device _as-is_.

Each device has it's own definition of management data format, which is completely device-specific. User needs to refer to the EP device manual for the data format and definitions. 

This command looks like a development-phase feature that engineers use, but the APPS team wants to keep it. Otherwise we would have dropped this command in final release.

We do need to print a warning message to make user aware that the command could cause possible issues with EP, if it is not used properly.

Reply:
Well, I don't like it at all. I'd much prefer actual commands for various EP devices be coded up instead of making users do it manually with hex numbers. That's way too ugly, error prone and time wasting; and will just make users write even uglier scripts to deal with the mess. Better to do it right than provide a horrid interface for people to do it wrong. 

Remember, also, this is open source so if you provide examples for common EPs you can let users add others that do not exist by following your example.